File: System/Workspace/Allors.Workspace.Protocol.Json/Data/ToJsonVisitor.cs
// <copyright file="ToJsonVisitor.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Protocol.Json
{
    using System.Collections.Generic;
    using System.Linq;
    using Allors.Protocol.Json;
    using Allors.Protocol.Json.Data;
    using Data;
    using Meta;
    using Extent = Allors.Protocol.Json.Data.Extent;
    using IVisitor = Data.IVisitor;
    using Node = Allors.Protocol.Json.Data.Node;
    using Procedure = Allors.Protocol.Json.Data.Procedure;
    using Pull = Allors.Protocol.Json.Data.Pull;
    using Result = Allors.Protocol.Json.Data.Result;
    using Select = Allors.Protocol.Json.Data.Select;
    using Sort = Allors.Protocol.Json.Data.Sort;
    using Step = Allors.Protocol.Json.Data.Step;

    public class ToJsonVisitor : IVisitor
    {
        private readonly Stack<Extent> extents;
        private readonly Stack<Predicate> predicates;
        private readonly Stack<Result> results;
        private readonly Stack<Select> selects;
        private readonly Stack<Step> steps;
        private readonly Stack<Node> nodes;
        private readonly Stack<Sort> sorts;

        public ToJsonVisitor()
        {
            this.extents = new Stack<Extent>();
            this.predicates = new Stack<Predicate>();
            this.results = new Stack<Result>();
            this.selects = new Stack<Select>();
            this.steps = new Stack<Step>();
            this.nodes = new Stack<Node>();
            this.sorts = new Stack<Sort>();
        }

        public Pull Pull { get; private set; }

        public Extent Extent => this.extents?.Peek();

        public Select Select => this.selects?.Peek();

        public Procedure Procedure { get; private set; }

        public void VisitAnd(And visited)
        {
            var predicate = new Predicate
            {
                Kind = PredicateKind.And,
                Dependencies = visited.Dependencies,
            };

            this.predicates.Push(predicate);

            if (visited.Operands != null)
            {
                var length = visited.Operands.Length;
                predicate.Operands = new Predicate[length];
                for (var i = 0; i < length; i++)
                {
                    var operand = visited.Operands[i];
                    operand.Accept(this);
                    predicate.Operands[i] = this.predicates.Pop();
                }
            }
        }

        public void VisitBetween(Between visited)
        {
            var predicate = new Predicate
            {
                Kind = PredicateKind.Between,
                Dependencies = visited.Dependencies,
                RoleType = visited.RoleType?.RelationType.Id,
                Values = visited.Values.Select(UnitConvert.ToString).ToArray(),
                Parameter = visited.Parameter,
            };

            this.predicates.Push(predicate);
        }

        public void VisitContainedIn(ContainedIn visited)
        {
            var predicate = new Predicate
            {
                Kind = PredicateKind.ContainedIn,
                Dependencies = visited.Dependencies,
                AssociationType = (visited.PropertyType as IAssociationType)?.RelationType.Id,
                RoleType = (visited.PropertyType as IRoleType)?.RelationType.Id,
                Values = visited.Objects?.Select(v => v.Identity.ToString()).ToArray(),
                Parameter = visited.Parameter,
            };

            this.predicates.Push(predicate);

            if (visited.Extent != null)
            {
                visited.Extent.Accept(this);
                predicate.Extent = this.extents.Pop();
            }
        }

        public void VisitContains(Contains visited)
        {
            var predicate = new Predicate
            {
                Kind = PredicateKind.Contains,
                Dependencies = visited.Dependencies,
                AssociationType = (visited.PropertyType as IAssociationType)?.RelationType.Id,
                RoleType = (visited.PropertyType as IRoleType)?.RelationType.Id,
                Object = visited.Object?.Identity.ToString(),
                Parameter = visited.Parameter,
            };

            this.predicates.Push(predicate);
        }

        public void VisitEquals(Equals visited)
        {
            var predicate = new Predicate
            {
                Kind = PredicateKind.Equals,
                Dependencies = visited.Dependencies,
                AssociationType = (visited.PropertyType as IAssociationType)?.RelationType.Id,
                RoleType = (visited.PropertyType as IRoleType)?.RelationType.Id,
                Object = visited.Object?.Identity.ToString(),
                Value = UnitConvert.ToString(visited.Value),
                Parameter = visited.Parameter,
            };

            this.predicates.Push(predicate);
        }

        public void VisitExcept(Except visited)
        {
            var extent = new Extent
            {
                Kind = ExtentKind.Except,
            };

            this.extents.Push(extent);

            if (visited.Operands != null)
            {
                var length = visited.Operands.Length;
                extent.Operands = new Extent[length];
                for (var i = 0; i < length; i++)
                {
                    var operand = visited.Operands[i];
                    operand.Accept(this);
                    extent.Operands[i] = this.extents.Pop();
                }
            }

            if (visited.Sorting?.Length > 0)
            {
                var length = visited.Sorting.Length;
                extent.Sorting = new Sort[length];
                for (var i = 0; i < length; i++)
                {
                    var sorting = visited.Sorting[i];
                    sorting.Accept(this);
                    extent.Sorting[i] = this.sorts.Pop();
                }
            }
        }

        public void VisitExists(Exists visited)
        {
            var predicate = new Predicate
            {
                Kind = PredicateKind.Exists,
                Dependencies = visited.Dependencies,
                AssociationType = (visited.PropertyType as IAssociationType)?.RelationType.Id,
                RoleType = (visited.PropertyType as IRoleType)?.RelationType.Id,
                Parameter = visited.Parameter,
            };

            this.predicates.Push(predicate);
        }

        public void VisitExtent(Allors.Workspace.Data.Extent visited)
        {
            var extent = new Extent
            {
                Kind = ExtentKind.Extent,
                ObjectType = visited.ObjectType?.Id,
                Sorting = visited.Sorting?.Select(v => new Sort { Descending = v.Descending, RoleType = v.RoleType?.RelationType.Id }).ToArray(),
            };

            this.extents.Push(extent);

            if (visited.Predicate != null)
            {
                visited.Predicate.Accept(this);
                extent.Predicate = this.predicates.Pop();
            }
        }

        public void VisitSelect(Allors.Workspace.Data.Select visited)
        {
            var @select = new Select();

            this.selects.Push(@select);

            if (visited.Step != null)
            {
                visited.Step.Accept(this);
                @select.Step = this.steps.Pop();
            }

            if (visited.Include?.Length > 0)
            {
                var length = visited.Include.Length;
                @select.Include = new Node[length];
                for (var i = 0; i < length; i++)
                {
                    var node = visited.Include[i];
                    node.Accept(this);
                    @select.Include[i] = this.nodes.Pop();
                }
            }
        }

        public void VisitGreaterThan(GreaterThan visited)
        {
            var predicate = new Predicate
            {
                Kind = PredicateKind.GreaterThan,
                Dependencies = visited.Dependencies,
                RoleType = visited.RoleType?.RelationType.Id,
                Value = UnitConvert.ToString(visited.Value),
                Parameter = visited.Parameter,
            };

            this.predicates.Push(predicate);
        }

        public void VisitInstanceOf(Instanceof visited)
        {
            var predicate = new Predicate
            {
                Kind = PredicateKind.InstanceOf,
                Dependencies = visited.Dependencies,
                ObjectType = visited.ObjectType?.Id,
                AssociationType = (visited.PropertyType as IAssociationType)?.RelationType.Id,
                RoleType = (visited.PropertyType as IRoleType)?.RelationType.Id,
            };

            this.predicates.Push(predicate);
        }

        public void VisitIntersect(Intersect visited)
        {
            var extent = new Extent
            {
                Kind = ExtentKind.Intersect,
            };

            this.extents.Push(extent);

            if (visited.Operands != null)
            {
                var length = visited.Operands.Length;
                extent.Operands = new Extent[length];
                for (var i = 0; i < length; i++)
                {
                    var operand = visited.Operands[i];
                    operand.Accept(this);
                    extent.Operands[i] = this.extents.Pop();
                }
            }

            if (visited.Sorting?.Length > 0)
            {
                var length = visited.Sorting.Length;
                extent.Sorting = new Sort[length];
                for (var i = 0; i < length; i++)
                {
                    var sorting = visited.Sorting[i];
                    sorting.Accept(this);
                    extent.Sorting[i] = this.sorts.Pop();
                }
            }
        }

        public void VisitLessThan(LessThan visited)
        {
            var predicate = new Predicate
            {
                Kind = PredicateKind.LessThan,
                Dependencies = visited.Dependencies,
                RoleType = visited.RoleType?.RelationType.Id,
                Value = UnitConvert.ToString(visited.Value),
                Parameter = visited.Parameter,
            };

            this.predicates.Push(predicate);
        }

        public void VisitLike(Like visited)
        {
            var predicate = new Predicate
            {
                Kind = PredicateKind.Like,
                Dependencies = visited.Dependencies,
                RoleType = visited.RoleType?.RelationType.Id,
                Value = UnitConvert.ToString(visited.Value),
                Parameter = visited.Parameter,
            };

            this.predicates.Push(predicate);
        }

        public void VisitNode(Allors.Workspace.Data.Node visited)
        {
            var node = new Node
            {
                AssociationType = (visited.PropertyType as IAssociationType)?.RelationType.Id,
                RoleType = (visited.PropertyType as IRoleType)?.RelationType.Id,
            };

            this.nodes.Push(node);

            if (visited.Nodes?.Length > 0)
            {
                var length = visited.Nodes.Length;
                node.Nodes = new Node[length];
                for (var i = 0; i < length; i++)
                {
                    visited.Nodes[i].Accept(this);
                    node.Nodes[i] = this.nodes.Pop();
                }
            }
        }

        public void VisitNot(Not visited)
        {
            var predicate = new Predicate()
            {
                Kind = PredicateKind.Not,
                Dependencies = visited.Dependencies,
            };

            this.predicates.Push(predicate);

            if (visited.Operand != null)
            {
                visited.Operand.Accept(this);
                predicate.Operand = this.predicates.Pop();
            }
        }

        public void VisitOr(Or visited)
        {
            var predicate = new Predicate()
            {
                Kind = PredicateKind.Or,
                Dependencies = visited.Dependencies,
            };

            this.predicates.Push(predicate);

            if (visited.Operands != null)
            {
                var length = visited.Operands.Length;
                predicate.Operands = new Predicate[length];
                for (var i = 0; i < length; i++)
                {
                    var operand = visited.Operands[i];
                    operand.Accept(this);
                    predicate.Operands[i] = this.predicates.Pop();
                }
            }
        }

        public void VisitPull(Allors.Workspace.Data.Pull visited)
        {
            var pull = new Pull
            {
                ExtentRef = visited.ExtentRef,
                ObjectType = visited.ObjectType?.Id,
                Object = visited.ObjectId ?? visited.Object?.Identity.ToString(),
            };

            if (visited.Extent != null)
            {
                visited.Extent.Accept(this);
                pull.Extent = this.extents.Pop();
            }

            if (visited.Results?.Length > 0)
            {
                var length = visited.Results.Length;

                pull.Results = new Result[length];
                for (var i = 0; i < length; i++)
                {
                    var result = visited.Results[i];
                    result.Accept(this);
                    pull.Results[i] = this.results.Pop();
                }
            }

            this.Pull = pull;
        }

        public void VisitResult(Allors.Workspace.Data.Result visited)
        {
            var result = new Result
            {
                SelectRef = visited.SelectRef,
                Name = visited.Name,
                Skip = visited.Skip,
                Take = visited.Take,
            };

            this.results.Push(result);

            if (visited.Select != null)
            {
                visited.Select.Accept(this);
                result.Select = this.selects.Pop();
            }
        }

        public void VisitSort(Allors.Workspace.Data.Sort visited)
        {
            var sort = new Sort
            {
                Descending = visited.Descending,
                RoleType = visited.RoleType?.RelationType.Id,
            };

            this.sorts.Push(sort);
        }

        public void VisitStep(Allors.Workspace.Data.Step visited)
        {
            var step = new Step
            {
                AssociationType = (visited.PropertyType as IAssociationType)?.RelationType.Id,
                RoleType = (visited.PropertyType as IRoleType)?.RelationType.Id,
            };

            this.steps.Push(step);

            if (visited.Include?.Length > 0)
            {
                var length = visited.Include.Length;
                step.Include = new Node[length];
                for (var i = 0; i < length; i++)
                {
                    var node = visited.Include[i];
                    node.Accept(this);
                    step.Include[i] = this.nodes.Pop();
                }
            }

            if (visited.Next != null)
            {
                visited.Next.Accept(this);
                step.Next = this.steps.Pop();
            }
        }

        public void VisitUnion(Union visited)
        {
            var extent = new Extent
            {
                Kind = ExtentKind.Union,
            };

            this.extents.Push(extent);

            if (visited.Operands != null)
            {
                var length = visited.Operands.Length;
                extent.Operands = new Extent[length];
                for (var i = 0; i < length; i++)
                {
                    var operand = visited.Operands[i];
                    operand.Accept(this);
                    extent.Operands[i] = this.extents.Pop();
                }
            }

            if (visited.Sorting?.Length > 0)
            {
                var length = visited.Sorting.Length;
                extent.Sorting = new Sort[length];
                for (var i = 0; i < length; i++)
                {
                    var sorting = visited.Sorting[i];
                    sorting.Accept(this);
                    extent.Sorting[i] = this.sorts.Pop();
                }
            }
        }

        public void VisitProcedure(Data.Procedure procedure) => this.Procedure = new Allors.Protocol.Json.Data.Procedure
        {
            Name = procedure.Name,
            CollectionByName = procedure.CollectionByName.ToJsonForCollectionByName(),
            ObjectByName = procedure.ObjectByName.ToJsonForObjectByName(),
            ValueByName = procedure.ValueByName.ToJsonForValueByName(),
            VersionByObject = procedure.VersionByObject.ToJsonForVersionByObject(),
        };
    }
}

File: typescript/libs/protocol/json/system/src/lib/ToJsonVisitor.ts
import { IVisitor, And, Between, ContainedIn, Contains, Equals, Except, Exists, GreaterThan, Instanceof, Intersect, LessThan, Like, Not, Or, Union } from '@allors/workspace/domain/system';
import { Extent as DataExtent, Select as DataSelect, Node as DataNode, Pull as DataPull, Result as DataResult, Sort as DataSort, Step as DataStep, Procedure as DataProcedure } from '@allors/workspace/domain/system';
import { AssociationType, RoleType } from '@allors/workspace/meta/system';
import { Extent } from './data/Extent';
import { ExtentKind } from './data/ExtentKind';
import { Node } from './data/Node';
import { Predicate } from './data/Predicate';
import { PredicateKind } from './data/PredicateKind';
import { Procedure } from './data/Procedure';
import { Pull } from './data/Pull';
import { Result } from './data/Result';
import { Select } from './data/Select';
import { Sort } from './data/Sort';
import { Step } from './data/Step';
import { UnitConvert } from './UnitConvert';

export class ToJsonVisitor implements IVisitor {
  private extents: Extent[];

  private predicates: Predicate[];

  private results: Result[];

  private selects: Select[];

  private steps: Step[];

  private nodes: Node[];

  private sorts: Sort[];

  Pull: Pull;

  Procedure: Procedure;

  get Extent(): Extent {
    return this.extents.pop();
  }

  get Select(): Select {
    return this.selects.pop();
  }

  public constructor(private unitConvert: UnitConvert) {
    this.extents = [];
    this.predicates = [];
    this.results = [];
    this.selects = [];
    this.steps = [];
    this.nodes = [];
    this.sorts = [];
  }

  public VisitAnd(visited: And) {
    const predicate: Predicate = {
      k: PredicateKind.And,
      d: visited.dependencies,
    };

    this.predicates.push(predicate);

    if (visited.operands != null) {
      const length = visited.operands.length;
      predicate.ops = [];
      for (let i = 0; i < length; i++) {
        const operand = visited.operands[i];
        operand.accept(this);
        predicate.ops[i] = this.predicates.pop();
      }
    }
  }

  public VisitBetween(visited: Between) {
    const predicate: Predicate = {
      k: PredicateKind.Between,
      d: visited.dependencies,
      r: visited.roleType?.relationType.tag,
      vs: visited.values?.map((v) => this.unitConvert.ToJson(v)),
      pas: visited.paths?.map((v) => v.relationType.tag),
      p: visited.parameter,
    };

    this.predicates.push(predicate);
  }

  public VisitContainedIn(visited: ContainedIn) {
    const predicate: Predicate = {
      k: PredicateKind.ContainedIn,
      d: visited.dependencies,
      a: (visited.propertyType as AssociationType)?.relationType.tag,
      r: (visited.propertyType as RoleType)?.relationType.tag,
      vs: visited.objects?.map((v) => v.id),
      p: visited.parameter,
    };

    this.predicates.push(predicate);

    if (visited.extent != null) {
      visited.extent.accept(this);
      predicate.e = this.extents.pop();
    }
  }

  public VisitContains(visited: Contains) {
    const predicate: Predicate = {
      k: PredicateKind.Contains,
      d: visited.dependencies,
      a: visited.propertyType.isAssociationType ? (visited.propertyType as AssociationType)?.relationType.tag : undefined,
      r: visited.propertyType.isRoleType ? (visited.propertyType as RoleType)?.relationType.tag : undefined,
      ob: visited.object?.id,
      p: visited.parameter,
    };

    this.predicates.push(predicate);
  }

  public VisitEquals(visited: Equals) {
    const predicate: Predicate = {
      k: PredicateKind.Equals,
      d: visited.dependencies,
      a: visited.propertyType.isAssociationType ? (visited.propertyType as AssociationType)?.relationType.tag : undefined,
      r: visited.propertyType.isRoleType ? (visited.propertyType as RoleType)?.relationType.tag : undefined,
      ob: visited.object?.id,
      v: this.unitConvert.ToJson(visited.value),
      pa: visited.path?.relationType.tag,
      p: visited.parameter,
    };

    this.predicates.push(predicate);
  }

  public VisitExcept(visited: Except) {
    const extent: Extent = {
      k: ExtentKind.Except,
    };

    this.extents.push(extent);
    if (visited.operands != null) {
      const length = visited.operands.length;
      extent.o = [];
      for (let i = 0; i < length; i++) {
        const operand = visited.operands[i];
        operand.accept(this);
        extent.o[i] = this.extents.pop();
      }
    }

    if (visited.sorting?.length > 0) {
      const length = visited.sorting.length;
      extent.s = [];
      for (let i = 0; i < length; i++) {
        const sorting = visited.sorting[i];
        sorting.accept(this);
        extent.s[i] = this.sorts.pop();
      }
    }
  }

  public VisitExists(visited: Exists) {
    const predicate: Predicate = {
      k: PredicateKind.Exists,
      d: visited.dependencies,
      a: visited.propertyType.isAssociationType ? (visited.propertyType as AssociationType)?.relationType.tag : undefined,
      r: visited.propertyType.isRoleType ? (visited.propertyType as RoleType)?.relationType.tag : undefined,
      p: visited.parameter,
    };

    this.predicates.push(predicate);
  }

  public VisitExtent(visited: DataExtent) {
    const extent: Extent = {
      k: ExtentKind.Extent,
      t: visited.objectType?.tag,
      s: visited.sorting?.map((v) => ({
        d: v.sortDirection,
        r: v.roleType?.relationType.tag,
      })),
    };

    this.extents.push(extent);
    if (visited.predicate != null) {
      visited.predicate.accept(this);
      extent.p = this.predicates.pop();
    }
  }

  public VisitSelect(visited: DataSelect) {
    const select: Select = {} as any;

    this.selects.push(select);

    if (visited.step != null) {
      visited.step.accept(this);
      select.s = this.steps.pop();
    }

    if (visited.include?.length > 0) {
      const includes = visited.include;
      const length = includes.length;
      select.i = [];
      for (let i = 0; i < length; i++) {
        const node = includes[i];
        node.accept(this);
        select.i[i] = this.nodes.pop();
      }
    }
  }

  public VisitGreaterThan(visited: GreaterThan) {
    const predicate: Predicate = {
      k: PredicateKind.GreaterThan,
      d: visited.dependencies,
      r: visited.roleType?.relationType.tag,
      v: this.unitConvert.ToJson(visited.value),
      pa: visited.path?.relationType.tag,
      p: visited.parameter,
    };

    this.predicates.push(predicate);
  }

  public VisitInstanceOf(visited: Instanceof) {
    const predicate: Predicate = {
      k: PredicateKind.InstanceOf,
      d: visited.dependencies,
      o: visited.objectType?.tag,
      a: visited.propertyType.isAssociationType ? (visited.propertyType as AssociationType)?.relationType.tag : undefined,
      r: visited.propertyType.isRoleType ? (visited.propertyType as RoleType)?.relationType.tag : undefined,
    };

    this.predicates.push(predicate);
  }

  public VisitIntersect(visited: Intersect) {
    const extent: Extent = {
      k: ExtentKind.Intersect,
    };

    this.extents.push(extent);

    if (visited.operands != null) {
      const length = visited.operands.length;
      extent.o = [];
      for (let i = 0; i < length; i++) {
        const operand = visited.operands[i];
        operand.accept(this);
        extent.o[i] = this.extents.pop();
      }
    }

    if (visited.sorting?.length > 0) {
      const length = visited.sorting.length;
      extent.s = [];
      for (let i = 0; i < length; i++) {
        const sorting = visited.sorting[i];
        sorting.accept(this);
        extent.s[i] = this.sorts.pop();
      }
    }
  }

  public VisitLessThan(visited: LessThan) {
    const predicate: Predicate = {
      k: PredicateKind.LessThan,
      d: visited.dependencies,
      r: visited.roleType?.relationType.tag,
      v: this.unitConvert.ToJson(visited.value),
      pa: visited.path?.relationType.tag,
      p: visited.parameter,
    };

    this.predicates.push(predicate);
  }

  public VisitLike(visited: Like) {
    const predicate: Predicate = {
      k: PredicateKind.Like,
      d: visited.dependencies,
      r: visited.roleType?.relationType.tag,
      v: this.unitConvert.ToJson(visited.value),
      p: visited.parameter,
    };

    this.predicates.push(predicate);
  }

  public VisitNode(visited: DataNode) {
    const node: Node = {
      a: visited.propertyType.isAssociationType ? (visited.propertyType as AssociationType)?.relationType.tag : undefined,
      r: visited.propertyType.isRoleType ? (visited.propertyType as RoleType)?.relationType.tag : undefined,
    };

    this.nodes.push(node);
    if (visited.nodes?.length > 0) {
      const length = visited.nodes.length;
      node.n = [];
      for (let i = 0; i < length; i++) {
        visited.nodes[i].accept(this);
        node.n[i] = this.nodes.pop();
      }
    }
  }

  public VisitNot(visited: Not) {
    const predicate: Predicate = {
      k: PredicateKind.Not,
      d: visited.dependencies,
    };
    this.predicates.push(predicate);
    if (visited.operand != null) {
      visited.operand.accept(this);
      predicate.op = this.predicates.pop();
    }
  }

  public VisitOr(visited: Or) {
    const predicate: Predicate = {
      k: PredicateKind.Or,
      d: visited.dependencies,
    };

    this.predicates.push(predicate);
    if (visited.operands != null) {
      const length = visited.operands.length;
      predicate.ops = [];
      for (let i = 0; i < length; i++) {
        const operand = visited.operands[i];
        operand.accept(this);
        predicate.ops[i] = this.predicates.pop();
      }
    }
  }

  public VisitPull(visited: DataPull) {
    const pull: Pull = {
      er: visited.extentRef,
      t: visited.objectType?.tag,
      o: visited.objectId ?? visited.object?.id,
    };

    if (visited.extent != null) {
      visited.extent.accept(this);
      pull.e = this.extents.pop();
    }

    if (visited.results?.length > 0) {
      const length = visited.results.length;
      pull.r = [];
      for (let i = 0; i < length; i++) {
        const result = visited.results[i];
        result.accept(this);
        pull.r[i] = this.results.pop();
      }
    }

    this.Pull = pull;
  }

  public VisitResult(visited: DataResult) {
    const result: Result = {
      r: visited.selectRef,
      n: visited.name,
      k: visited.skip,
      t: visited.take,
    };

    this.results.push(result);
    if (visited.select != null) {
      visited.select.accept(this);
      result.s = this.selects.pop();
    }
  }

  public VisitSort(visited: DataSort) {
    const sort: Sort = {
      d: visited.sortDirection,
      r: visited.roleType?.relationType.tag,
    };

    this.sorts.push(sort);
  }

  public VisitStep(visited: DataStep) {
    const step: Step = {
      a: visited.propertyType.isAssociationType ? (visited.propertyType as AssociationType)?.relationType.tag : undefined,
      r: visited.propertyType.isRoleType ? (visited.propertyType as RoleType)?.relationType.tag : undefined,
    };

    this.steps.push(step);
    if (visited.include?.length > 0) {
      const length = visited.include.length;
      step.i = [];
      for (let i = 0; i < length; i++) {
        const node = visited.include[i];
        node.accept(this);
        step.i[i] = this.nodes.pop();
      }
    }

    if (visited.next != null) {
      visited.next.accept(this);
      step.n = this.steps.pop();
    }
  }

  public VisitUnion(visited: Union) {
    const extent: Extent = {
      k: ExtentKind.Union,
    };

    this.extents.push(extent);

    if (visited.operands != null) {
      const length = visited.operands.length;
      extent.o = [];
      for (let i = 0; i < length; i++) {
        const operand = visited.operands[i];
        operand.accept(this);
        extent.o[i] = this.extents.pop();
      }
    }

    if (visited.sorting?.length > 0) {
      const length = visited.sorting.length;
      extent.s = [];
      for (let i = 0; i < length; i++) {
        const sorting = visited.sorting[i];
        sorting.accept(this);
        extent.s[i] = this.sorts.pop();
      }
    }
  }

  public VisitProcedure(procedure: DataProcedure) {
    const c = {};
    for (const [name, objects] of procedure.collections) {
      c[name] = objects;
    }

    const o = {};
    for (const [name, object] of procedure.objects) {
      o[name] = object;
    }

    const v = {};
    for (const [name, value] of procedure.values) {
      v[name] = value;
    }

    const p = [];
    for (const [object, version] of procedure.pool) {
      p.push([object.id, version]);
    }

    this.Procedure = {
      n: procedure.name,
      c,
      o,
      v,
      p,
    };
  }
}
